sips/sip-023/sip-023-emergency-fix-traits.md

```diff
@@ -0,0 +1,116 @@
+# Preamble
+
+SIP Number: 023
+
+Title: Emergency Fix to Trait Invocation Behavior
+
+Authors:
+    Aaron Blankstein <aaron@hiro.so>,
+    Brice Dobry <brice@hiro.so>,
+    Jude Nelson <jude@stacks.org>,
+
+Consideration: Technical, Governance
+
+Type: Consensus
+
+Status: Draft
+
+Created: 1 May 2023
+
+License: BSD 2-Clause
+
+Sign-off: 
+
+Discussions-To: https://github.com/stacksgov/sips
+
+# Abstract
+
+On 1 May 2023, it was discovered that pre-2.1 contracts exposing public methods with
+trait arguments could not be invoked with previously working trait-implementing
+contract arguments.
+
+This bug was caused by the activation of Stacks Epoch 2.2 (https://github.com/stacksgov/sips/blob/main/sips/sip-022/sip-022-emergency-pox-fix.md).
+
+This SIP proposes an **immediate consensus-breaking change** to
+introduce a new Stacks Epoch 2.3 that corrects this regression.
+
+**This SIP proposes a Bitcoin activation height of 788,287**
+
+# Epoch 2.2 Bug Behavior
+
+Stacks Epoch 2.1 introduced a new type checker and type system which
+impacts trait invocations. In order for existing contracts to remain
+compatible, their types must be _canonicalized_. The canonicalization
+method performed an exact check for the current epoch:
+
+```rust
+    pub fn canonicalize(&self, epoch: &StacksEpochId) -> TypeSignature {
+        match epoch {
+            StacksEpochId::Epoch21 => self.canonicalize_v2_1(),
+            _ => self.clone(),
+        }
+    }
+```
+
+Therefore, a pre-2.1 function with trait arguments that is invoked in Epoch 2.2
+will fail to canonicalize its trait arguments, and abort with a
+runtime error. Specifically, if a miner includes a contract call transaction in a block, it will be rejected by the mempool
+with a type error, and a read-only call will fail with a runtime error.
+
+# Specification
+
+The Epoch 2.3 hard fork will do the following:
+
+* Update the `canonicalize()` method to match on all epochs, setting
+  the Epoch 2.3 behavior to `self.canonicalize_v2_1()`, and the Epoch
+  2.2 behavior to `self.clone()`.
+  
+This will preserve the buggy 2.2 behavior during the 2.2 epoch (so that the
+hard fork does not require rollback), but fix the behavior after activation
+of the 2.3 epoch.
+
+# Related Work
```

Comment:
Can a brief note be added to this section about why the attempted workarounds were insufficient?

Reply:
Here's some potential content for this:

Potential workarounds were explored first to try to solve this issue without a hard-fork. Unfortunately attempts to wrap pre-2.1 contracts with 2.2 contracts can avoid the mempool rejection, but still hit the same error in the form of a runtime type-checker error. Upon further inspection into the code paths, the hard-fork was determined to be the only option in this case.